File: inventory-service/src/main/java/com/catius/inventory/service/InventoryService.java
package com.catius.inventory.service;

import com.catius.inventory.domain.Inventory;
import com.catius.inventory.domain.exception.ProductNotFoundException;
import com.catius.inventory.repository.InventoryRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class InventoryService {

    private final InventoryRepository repository;

    public Inventory getInventory(Long productId) {
        return repository.findByProductId(productId)
                .orElseThrow(() -> new ProductNotFoundException(productId));
    }

    @Transactional
    public void reserve(Long productId, int quantity) {
        Inventory inventory = repository.findByProductIdForUpdate(productId)
                .orElseThrow(() -> new ProductNotFoundException(productId));
        inventory.reserve(quantity);
    }

    @Transactional
    public void release(Long productId, int quantity) {
        Inventory inventory = repository.findByProductIdForUpdate(productId)
                .orElseThrow(() -> new ProductNotFoundException(productId));
        inventory.release(quantity);
    }
}

File: inventory-service/src/main/kotlin/com/catius/inventory/service/InventoryService.kt
package com.catius.inventory.service

import com.catius.inventory.domain.MovementType
import com.catius.inventory.domain.StockMovement
import com.catius.inventory.domain.exception.InventoryNotFoundException
import com.catius.inventory.domain.port.InventoryRepository
import com.catius.inventory.domain.port.StockMovementRepository
import com.catius.inventory.domain.usecase.FindInventoryUseCase
import com.catius.inventory.domain.usecase.ReleaseInventoryUseCase
import com.catius.inventory.domain.usecase.ReserveInventoryUseCase
import com.catius.inventory.domain.usecase.command.InventoryCommand
import com.catius.inventory.domain.usecase.command.InventoryResult
import com.catius.inventory.domain.usecase.command.ReserveResult
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.Clock
import java.time.Instant

@Service
class InventoryService(
    private val inventoryRepository: InventoryRepository,
    private val stockMovementApplier: StockMovementApplier,
    private val clock: Clock,
) : ReserveInventoryUseCase, ReleaseInventoryUseCase, FindInventoryUseCase {

    @Transactional
    override fun reserve(command: InventoryCommand.ReserveCommand): ReserveResult {
        val now = Instant.now(clock)
        command.items.forEach { item ->
            stockMovementApplier.apply(
                orderId = command.orderId,
                productId = item.productId,
                quantity = item.quantity,
                type = MovementType.RESERVE,
                now = now,
            )
        }

        return ReserveResult(
            orderId = command.orderId,
            items = command.items.map {
                ReserveResult.Item(productId = it.productId, quantity = it.quantity)
            },
        )
    }

    @Transactional
    override fun release(command: InventoryCommand.ReleaseCommand) {
        val now = Instant.now(clock)
        command.items.forEach { item ->
            stockMovementApplier.apply(
                orderId = command.orderId,
                productId = item.productId,
                quantity = item.quantity,
                type = MovementType.RELEASE,
                now = now,
            )
        }
    }

    @Transactional(readOnly = true)
    override fun findByProductId(productId: Long): InventoryResult {
        val inventory = inventoryRepository.findByProductId(productId)
            ?: throw InventoryNotFoundException(productId)
        return InventoryResult(
            productId = inventory.productId,
            availableQuantity = inventory.availableQuantity,
            updatedAt = inventory.updatedAt,
        )
    }
}
